### PR TITLE
Deploy history

### PR DESCRIPTION
**Pull request by byroot**
Fixes: https://github.com/Shopify/shipit-engine/issues/414

This is work in progress but I'm about to onboard.

capture d ecran 2015-06-01 a 21 03 41

capture d ecran 2015-06-01 a 21 03 49

- I removed the rollbacks button because I don't think it's a good idea to rollback that far. But If you guys think I should let them let me know.
- Since I reuse the API paginator I'm limited to `first` and `next` links. We can always improve that latter.
- [x] Needs tests

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -12,7 +12,7 @@ PATH
       jquery-rails (~> 4.0.3)
       octokit (~> 3.8.0)
       omniauth-github (~> 1.1.2)
-      pubsubstub (~> 0.0.11)
+      pubsubstub (~> 0.0.9)
```
Comment (rafaelfranca) [inline Gemfile.lock]: Any specific reason for the downgrade?

Comment (byroot) [inline Gemfile.lock]: It's a mistake /o\

File: app/controllers/concerns/paginator.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
+class Paginator
```
Comment (rafaelfranca) [inline app/controllers/concerns/paginator.rb]: Maybe this class should be moved to `lib`. It is used on controllers but it is not a concern related just to the controller layer. Also, there is already a lot of logic here to justify unit tests for it.

File: config/routes.rb [hunk 3]
```diff
@@ -44,6 +44,7 @@
     resources :rollbacks, only: %i(create)
     resources :tasks, only: %i(show) do
       collection do
+        get '' => 'tasks#index', as: :index
```
Comment (rafaelfranca) [inline config/routes.rb]: I think you can change the `only` option to include `index` and get the same result.

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -1,6 +1,16 @@
 class TasksController < ShipitController
+  include Pagination
+
   before_action :stack
 
+  self.default_page_size = 20
+
+  def index
+    @paginator = paginate(stack.deploys)
```
Comment (rafaelfranca) [inline app/controllers/tasks_controller.rb]: This variable is not being used in any view so I think we can change it to be a local variable instead a instance variable.

File: app/views/tasks/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+<div class="wrapper"
+  <section>
+    <ul class="commit-list">
+      <%= render @tasks, read_only: true %>
+    </ul>
+  </section>
+</div>
+
+<div class="wrapper"
```
Comment (rafaelfranca) [inline app/views/tasks/index.html.erb]: Missing close `>`

File: app/views/tasks/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+<%= render partial: "stacks/header", locals: { stack: @stack } %>
+
+<div class="wrapper"
```
Comment (rafaelfranca) [inline app/views/tasks/index.html.erb]: Missing close `>`

File: app/views/tasks/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  <section>
+    <% if params[:since].present? %>
+      <%= link_to 'Start', @links[:first], class: 'header__btn btn' %>
+    <% else %>
+      <%= link_to 'Start', '', class: 'header__btn btn btn--disabled disabled' %>
+    <% end %>
+
+    <% if url = @links[:next] %>
```
Comment (rafaelfranca) [inline app/views/tasks/index.html.erb]: This line give a ruby warning. I hate this warning too but it is a good practice to write warning free code. To do so put the `url = @links[:next]` inside parentheses

**Comment by rafaelfranca** [comment]
Just a few small comments but the feature looks good.

**Comment by byroot** [comment]
All good points thanks!

File: app/views/tasks/index.html.erb (hunk 7 above)
Comment (byroot) [inline app/views/tasks/index.html.erb]: Not sure what the warning is supposed to be, I tried:

```ruby
@foo = {}

if url = @foo[:next]
  p url
end
```

And it didn't generate any warning.

Comment (rafaelfranca) [inline app/views/tasks/index.html.erb]: O_O I can't reproduce this warning anymore but I got it many times on the Rails repository. If I remember the situation I'll let you know but this is fine as it is.

**Comment by byroot** [comment]
I centered the navigation links. I believe this is good to go.

**Pull request merged**